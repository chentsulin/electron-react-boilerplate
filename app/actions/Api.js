// @flow
import type { counterStateType } from '../reducers/Api';

type actionType = {
  +type: string
};

export const INCREMENT_COUNTER = 'INCREMENT_COUNTER';
export const DECREMENT_COUNTER = 'DECREMENT_COUNTER';

export function increment() {
  return {
    type: INCREMENT_COUNTER
  };
}

export function decrement() {
  return {
    type: DECREMENT_COUNTER
  };
}

export function incrementIfOdd() {
  return (dispatch: (action: actionType) => void, getState: () => counterStateType) => {
    const { counter } = getState();

    if (counter % 2 === 0) {
      return;
    }

    dispatch(increment());
  };
}

export function incrementAsync(delay: number = 1000) {
  return (dispatch: (action: actionType) => void) => {
    setTimeout(() => {
      dispatch(increment());
    }, delay);
  };
}

// nicks new stuff
export function logResult(result) {
  console.log(result);
}

export function logError(error) {
  console.log('Looks like there was a problem: \n', error);
}

export function validateResponse(response) {
  if (!response.ok) {
    throw Error(response.statusText);
  }
  return response;
}

export function readResponseAsJSON(response) {
  return response.json();
}

export function fetchJSON(pathToResource) {
  fetch(pathToResource) // 1
  .then(validateResponse) // 2
  .then(readResponseAsJSON) // 3
  .then(logResult) // 4
  .catch(logError);
}
