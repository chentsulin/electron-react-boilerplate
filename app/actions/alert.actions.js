import { alertConstants } from '../constants';

export const alertActions = {
  success,
  error,
  clear
};

export default alertActions;

function success(message) {
  return { type: alertConstants.SUCCESS, message };
}

function error(message) {
  return { type: alertConstants.ERROR, message };
}

function clear() {
  return { type: alertConstants.CLEAR };
}
