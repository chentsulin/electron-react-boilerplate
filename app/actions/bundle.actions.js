import sort from 'fast-sort';
import traverse from 'traverse';
import { bundleConstants } from '../constants/bundle.constants';
import { bundleService } from '../services/bundle.service';

export const bundleActions = {
  mockFetchAll,
  fetchAll,
  delete: removeBundle,
  downloadResources,
  requestSaveBundleTo,
  toggleModePauseResume,
  toggleSelectBundle
};

export default bundleActions;

export function mockFetchAll() {
  return dispatch => {
    dispatch(request());
    return new Promise(resolve => {
      const mockBundles = getMockBundles();
      resolve(mockBundles);
    }).then(bundles => dispatch(success(bundles)), error => dispatch(failure(error)));
  };

  function request() {
    return { type: bundleConstants.FETCH_REQUEST };
  }
  function success(bundles) {
    return { type: bundleConstants.FETCH_SUCCESS, bundles };
  }
  function failure(error) {
    return { type: bundleConstants.FETCH_FAILURE, error };
  }
}

export function fetchAll() {
  return dispatch => {
    dispatch(request());

    return bundleService
      .fetchAll()
      .then(bundles => dispatch(success(bundles)), error => dispatch(failure(error)));
  };

  function request() {
    return { type: bundleConstants.FETCH_REQUEST };
  }
  function success(bundles) {
    return { type: bundleConstants.FETCH_SUCCESS, bundles };
  }
  function failure(error) {
    return { type: bundleConstants.FETCH_FAILURE, error };
  }
}

export function downloadResources(id) {
  return async dispatch => {
    try {
      const manifestResourcePaths = await bundleService.getManifestResourcePaths(id);
      manifestResourcePaths.unshift('metadata.xml');
      await bundleService.downloadResources(id);
      dispatch(request(id, manifestResourcePaths));
    } catch (error) {
      dispatch(failure(id, error));
    }
  };
  function request(_id, manifestResourcePaths) {
    return { type: bundleConstants.DOWNLOAD_RESOURCES_REQUEST, id: _id, manifestResourcePaths };
  }
  function failure(_id, error) {
    return { type: bundleConstants.DOWNLOAD_RESOURCES_FAILURE, id, error };
  }
}


function removeBundle(id) {
  return dispatch => {
    dispatch(request(id));

    bundleService
      .delete(id)
      .then(() => {
        dispatch(success(id));
        return true;
      })
      .catch(error => {
        dispatch(failure(id, error));
        return true;
      });
  };

  function request(_id) {
    return { type: bundleConstants.DELETE_REQUEST, id: _id };
  }
  function success(_id) {
    return { type: bundleConstants.DELETE_SUCCESS, id: _id };
  }
  function failure(_id, error) {
    return { type: bundleConstants.DELETE_FAILURE, id: _id, error };
  }
}

export function requestSaveBundleTo(id, selectedFolder) {
  return async dispatch => {
    const bundleInfo = await bundleService.fetchById(id);
    const bundleBytesToSave = traverse(bundleInfo.store.file_info).reduce(addByteSize, 0);
    const resourcePaths = await bundleService.getResourcePaths(id);
    resourcePaths.unshift('metadata.xml');
    const resourcePathsProgress = resourcePaths.reduce((acc, resourcePath) => {
      acc[resourcePath] = 0;
      return acc;
    }, {});
    let bundleBytesSaved = 0;
    dispatch(request(id, selectedFolder, bundleBytesToSave, resourcePaths));
    resourcePaths.forEach(async resourcePath => {
      try {
        const downloadItem = await bundleService.requestSaveResourceTo(
          selectedFolder,
          id,
          resourcePath,
          (resourceTotalBytesSaved, resourceProgress) => {
            const originalResourceBytesTransferred = resourcePathsProgress[resourcePath];
            resourcePathsProgress[resourcePath] = resourceTotalBytesSaved;
            const bytesDiff = resourceTotalBytesSaved - originalResourceBytesTransferred;
            bundleBytesSaved += bytesDiff;
            if (resourceProgress && resourceProgress % 100 === 0) {
              dispatch(updated(
                id,
                resourcePath,
                resourceTotalBytesSaved,
                bundleBytesSaved,
                bundleBytesToSave
              ));
            }
          }
        );
        return downloadItem;
      } catch (error) {
        dispatch(failure(id, error));
      }
    });
  };

  function addByteSize(accBytes, fileInfoNode) {
    if (fileInfoNode.is_dir || this.isRoot || fileInfoNode.size === undefined) {
      return accBytes;
    }
    return accBytes + fileInfoNode.size;
  }

  function request(_id, _folderName, bundleBytesToSave, resourcePaths) {
    return {
      type: bundleConstants.SAVETO_REQUEST,
      id: _id,
      folderName: _folderName,
      bundleBytesToSave,
      resourcePaths
    };
  }

  function updated(
    _id,
    resourcePath,
    resourceTotalBytesSaved,
    bundleBytesSaved,
    bundleBytesToSave
  ) {
    return {
      type: bundleConstants.SAVETO_UPDATED,
      id,
      resourcePath,
      resourceTotalBytesSaved,
      bundleBytesSaved,
      bundleBytesToSave
    };
  }
  function failure(_id, error) {
    return { type: bundleConstants.SAVETO_FAILURE, id: _id, error };
  }
}

export function toggleModePauseResume(id) {
  return { type: bundleConstants.TOGGLE_MODE_PAUSE_RESUME, id };
}

export function toggleSelectBundle(id) {
  return { type: bundleConstants.TOGGLE_SELECT, id };
}

function getMockBundles() {
  const bundles = [
    {
      id: 'bundle01',
      name: 'Test Bundle #1',
      revision: 3,
      task: 'UPLOAD',
      status: 'COMPLETED'
    },
    {
      id: 'bundle02',
      name: 'Another Bundle',
      revision: 3,
      task: 'UPLOAD',
      status: 'IN_PROGRESS',
      progress: 63,
      mode: 'PAUSED'
    },
    {
      id: 'bundle03',
      name: 'Audio Bundle',
      revision: 52,
      task: 'DOWNLOAD',
      status: 'IN_PROGRESS',
      progress: 12,
      mode: 'RUNNING'
    },
    {
      id: 'bundle04',
      name: 'Unfinished Bundle',
      task: 'UPLOAD',
      status: 'DRAFT'
    },
    {
      id: 'bundle05',
      name: 'Unfinished Video Bundle',
      task: 'UPLOAD',
      status: 'DRAFT'
    },
    {
      id: 'bundle06',
      name: 'DBL Bundle',
      task: 'DOWNLOAD',
      status: 'NOT_STARTED'
    },
    {
      id: 'bundle07',
      name: 'DBL Bundle 3',
      task: 'DOWNLOAD',
      status: 'NOT_STARTED'
    },
    {
      id: 'bundle08',
      name: 'Audio Bundle #2',
      revision: 40,
      task: 'DOWNLOAD',
      status: 'COMPLETED'
    },
    {
      id: 'bundle09',
      name: 'Audio Bundle #3',
      revision: 5,
      task: 'SAVETO',
      status: 'IN_PROGRESS',
      progress: 0,
    },
    {
      id: 'bundle10',
      name: 'Audio Bundle #4',
      revision: 4,
      task: 'SAVETO',
      status: 'IN_PROGRESS',
      progress: 66,
    },
    {
      id: 'bundle11',
      name: 'Audio Bundle #5',
      revision: 5,
      task: 'SAVETO',
      status: 'COMPLETED',
      progress: 100,
    }
  ];
  const taskOrder = ['UPLOAD', 'DOWNLOAD', 'SAVETO'];
  const statusOrder = ['IN_PROGRESS', 'DRAFT', 'COMPLETED', 'NOT_STARTED'];
  const sortedBundles = sort(bundles).asc([
    b => statusOrder.indexOf(b.status),
    b => taskOrder.indexOf(b.task),
    b => b.name
  ]);
  return sortedBundles;
}
