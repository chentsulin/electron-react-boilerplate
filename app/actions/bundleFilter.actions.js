import { bundleFilterConstants } from '../constants/bundleFilter.constants';

export const bundleFilterActions = {
  updateSearchInput,
  addSearchMatch,
  clearSearch
};

export default bundleFilterActions;

export function updateSearchInput(searchInput, bundles) {
  if (searchInput.length > 0) {
    return { type: bundleFilterConstants.UPDATE_SEARCH_INPUT, searchInput, bundles };
  }
  return clearSearch();
}

function addSearchMatch(bundle) {
  return { type: bundleFilterConstants.ADD_SEARCH_MATCH, bundle };
}

function clearSearch() {
  return { type: bundleFilterConstants.CLEAR_SEARCH_RESULTS };
}
