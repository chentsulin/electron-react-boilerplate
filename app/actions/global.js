/**
 * App actions
 */
export const SWITCH_VIEW = 'redux/Global/SWITCH_VIEW';
export const SET_GALLERY = 'redux/Global/SET_GALLERY';
export const LOCATION_CHANGE = '@@router/LOCATION_CHANGE';

/**
 * Dispatched when we want to set the current module and gallery ids.
 *
 * @param {string | number} moduleId - The module Id.
 * @param {string | number} galleryId - The gallery Id.
 * @param {string} searchQuery - The search query text.
 * @return {object} - An action object with a type of SET_GALLERY
 */
export function setGallery(moduleId, galleryId, searchQuery) {
  return {
    type: SET_GALLERY,
    payload: {
      moduleId,
      galleryId,
      searchQuery,
    },
  };
}

/**
 * Switch view form list to masonry
 *
 * @return {object} - An action object with a type of SWITCH_VIEW
 */
export function switchView() {
  return {
    type: SWITCH_VIEW,
  };
}
