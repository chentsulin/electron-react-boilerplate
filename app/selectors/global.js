/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectRoute = state => state.get('route');

const selectGlobal = state => state.get('global');

const makeSelectCurrentModuleId = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.get('currentModuleId'),
  );

const makeSelectCurrentGalleryId = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.get('currentGalleryId'),
  );

const makeSelectCurrentSearchQuery = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.get('currentSearchQuery'),
  );

const makeSelectModules = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.get('modules').toJS(),
  );

const makeSelectShowMasonry = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.get('showMasonry'),
  );

const makeSelectModule = moduleId =>
  createSelector(
    selectGlobal,
    globalState => globalState.getIn(['modules', moduleId]).toJS(),
  );

const makeSelectLocation = () =>
  createSelector(
    selectRoute,
    routeState => routeState.get('location').toJS(),
  );

const makeSelectRouteParams = () =>
  createSelector(
    selectRoute,
    () => {},
  );

export {
  makeSelectLocation,
  makeSelectModules,
  makeSelectModule,
  makeSelectRouteParams,
  makeSelectCurrentModuleId,
  makeSelectCurrentGalleryId,
  makeSelectCurrentSearchQuery,
  makeSelectShowMasonry,
};
