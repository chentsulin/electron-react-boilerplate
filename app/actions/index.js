import * as globalActions from './global';
import * as galleryActions from './gallery';
import * as languageActions from './language';

const actions = {
  ...globalActions,
  ...galleryActions,
  ...languageActions,
};

export default actions;
