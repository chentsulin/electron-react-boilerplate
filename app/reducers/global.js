/*
 * App reducer
 */
import { fromJS } from 'immutable';

import modules from '../modules';
import { SWITCH_VIEW, SET_GALLERY } from '../actions/global';

const initialStateBuilder = {
  modules: {},
  currentModuleId: null,
  currentGalleryId: null,
  currentSearchQuery: null,
  showMasonry: false,
};

// Load all of our modules
modules.keySeq().forEach(key => {
  initialStateBuilder.modules[key] = modules.get(key);
});

// The initial state of the App
export const initialState = fromJS(initialStateBuilder);

function globalReducer(state = initialState, action) {
  const { type, payload } = action;

  switch (type) {
    case SWITCH_VIEW: {
      return state.set('showMasonry', !state.get('showMasonry'));
    }
    case SET_GALLERY: {
      const { moduleId, galleryId, searchQuery } = payload;
      return state
        .set('currentModuleId', moduleId)
        .set('currentGalleryId', galleryId)
        .set('currentSearchQuery', searchQuery);
    }
    default:
      return state;
  }
}

export default globalReducer;
