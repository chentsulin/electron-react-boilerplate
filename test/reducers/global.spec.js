import Imutable from 'immutable';
import globalReducer from '../../app/reducers/gallery';
import { SET_GALLERY, SWITCH_VIEW } from '../../app/actions/global';

describe('reducers', () => {
  describe('gallery', () => {
    it('should handle initial state', () => {
      expect(globalReducer(undefined, {})).toMatchSnapshot();
    });

    it('should handle SET_GALLERY', () => {
      const state = Imutable.Map({});

      const action = {
        type: SET_GALLERY,
        payload: {
          moduleId: 'moduleId',
          galleryId: 'galleryId',
          searchQuery: null,
        },
      };

      expect(globalReducer(state, action)).toMatchSnapshot();
    });

    it('should handle SWITCH_VIEW', () => {
      const state = Imutable.Map({});

      const action = {
        type: SWITCH_VIEW,
      };

      expect(globalReducer(state, action)).toMatchSnapshot();
    });

    it('should handle unknown action type', () => {
      const state = Imutable.Map({});

      expect(globalReducer(state, { type: 'unknown' })).toMatchSnapshot();
    });
  });
});
