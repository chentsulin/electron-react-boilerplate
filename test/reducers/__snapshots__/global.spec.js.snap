// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers gallery should handle SET_GALLERY 1`] = `
Immutable.Map {
  "/gallery/moduleId/galleryId": Immutable.Map {
    "loading": false,
    "error": false,
    "hasNextPage": true,
    "pageNumber": 0,
    "items": Immutable.Map {},
    "itemsList": Immutable.List [],
    "scrollPosition": -1,
    "oldWidth": 0,
  },
}
`;

exports[`reducers gallery should handle SWITCH_VIEW 1`] = `Immutable.Map {}`;

exports[`reducers gallery should handle initial state 1`] = `
Immutable.Map {
  "lorumPicsumOne": Immutable.Map {
    "loading": false,
    "error": false,
    "hasNextPage": true,
    "pageNumber": 0,
    "items": Immutable.Map {},
    "itemsList": Immutable.List [],
    "scrollPosition": -1,
    "oldWidth": 0,
  },
  "lorumPicsumTwo": Immutable.Map {
    "loading": false,
    "error": false,
    "hasNextPage": true,
    "pageNumber": 0,
    "items": Immutable.Map {},
    "itemsList": Immutable.List [],
    "scrollPosition": -1,
    "oldWidth": 0,
  },
}
`;

exports[`reducers gallery should handle unknown action type 1`] = `Immutable.Map {}`;
